flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/OperatorState.java

```diff
@@ -118,14 +142,30 @@ public ByteStreamStateHandle getCoordinatorState() {
     }
 
     public Map<Integer, OperatorSubtaskState> getSubtaskStates() {
+        return getSubtaskStates(false);
+    }
+
+    public Map<Integer, OperatorSubtaskState> getSubtaskStates(boolean includeFinishedFlag) {
+        if (!includeFinishedFlag && isFullyFinished()) {
+            return Collections.emptyMap();
+        }
+
         return Collections.unmodifiableMap(operatorSubtaskStates);
     }
 
     public Collection<OperatorSubtaskState> getStates() {
+        if (isFullyFinished()) {
+            return Collections.emptySet();
+        }
```

Comment:
I'm not sure that this check is needed. If `isFullyFinished` then `subtaskStates` should be empty anyways. So continuing wouldn't hurt but the check makes me feel that it's **necessary** in **all** methods.

ditto: other methods

Probably a subclass `FinishedOperatorState` with always empty state would make it easier and less error-prone, WDYT?

Reply:
I think it would indeed make it more clear and not worry about consistency between methods. I updated the PR with this method. Very thanks!